tests/optional_kernel_features/kernel_features_common.h

```diff
@@ -0,0 +1,519 @@
+ *
+ * @tparam LambdaNoArg The type of lambda for single_task invocation
+ * @tparam LambdaItemArg The type of lambda for parallel_for invocation
+ * @tparam LambdaGroupArg The type of lambda for parallel_for_work_group
+ * invocation
+ * @param is_exception_expected The flag shows if exception expected from the
+ * kernel
+ * @param errc_expected The error code that expected from sycl::exception
+ * @param queue The sycl::queue instance for device
+ * @param separate_lambda_no_arg The lambda for single_task invocation
+ * @param separate_lambda_item_arg The lambda for parallel_for invocation
+ * @param separate_lambda_group_arg The lambda for parallel_for_work_group
+ * invocation
+ */
+template <typename LambdaNoArg, typename LambdaItemArg, typename LambdaGroupArg>
+void run_separate_lambda(const bool is_exception_expected,
+                         const sycl::errc errc_expected, sycl::queue &queue,
+                         LambdaNoArg separate_lambda_no_arg,
+                         LambdaItemArg separate_lambda_item_arg,
+                         LambdaGroupArg separate_lambda_group_arg) {
+  bool is_exception_thrown = false;
+  {
+    INFO("Execution of separate lambda function in single_task");
+    try {
+      queue
+          .submit([&](sycl::handler &cgh) {
+            cgh.single_task(separate_lambda_no_arg);
+          })
+          .wait();
+      is_exception_thrown = false;
+    } catch (const sycl::exception &e) {
+      is_exception_thrown = true;
+      if (is_exception_expected) {
+        INFO("Error code check");
+        CHECK(e.code() == errc_expected);
+      }
+    }
+    CHECK(is_exception_thrown == is_exception_expected);
+    check_async_exception(queue, false);
+  }
+  {
+    INFO("Execution of separate lambda function in parallel_for");
+    try {
+      queue
+          .submit([&](sycl::handler &cgh) {
+            cgh.parallel_for(sycl::range{1}, separate_lambda_item_arg);
+          })
+          .wait();
+      is_exception_thrown = false;
+    } catch (const sycl::exception &e) {
+      is_exception_thrown = true;
+      if (is_exception_expected) {
+        INFO("Error code check");
+        CHECK(e.code() == errc_expected);
+      }
+    }
+    CHECK(is_exception_thrown == is_exception_expected);
+    check_async_exception(queue, false);
+  }
+  {
+    INFO("Execution of separate lambda function in parallel_for_work_group");
+    try {
+      queue
+          .submit([&](sycl::handler &cgh) {
+            cgh.parallel_for_work_group(sycl::range{1}, sycl::range{1},
+                                        separate_lambda_group_arg);
+          })
+          .wait();
+      is_exception_thrown = false;
+    } catch (const sycl::exception &e) {
+      is_exception_thrown = true;
+      if (is_exception_expected) {
+        INFO("Error code check");
+        CHECK(e.code() == errc_expected);
+      }
+    }
+    CHECK(is_exception_thrown == is_exception_expected);
+    check_async_exception(queue, false);
```

Comment:
I have the feeling to mostly see 3 times the same code.
Factorize in a function/lambda called 3 times and taking a lambda for executing the different code?

Reply:
You're right, it looks like I can optimize these places. I tried to refactor these functions in this commit: 2fb2d5e4004329d31498a25c27355e0fc3c59dd7